### PR TITLE
Support DNS names for egress network policy

### PR DESCRIPTION
**Pull request by pravisankar**
Trello card: https://trello.com/c/yH9l0ouV/289-8-use-dns-names-with-egress-firewall-demo-human-names

### CODE REVIEW

**Comment by pravisankar** [comment]
@openshift/networking PTAL
(I think it will be easy to follow if you review one commit at a time)

File: pkg/sdn/plugin/egress_dns.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+		}
+
+		for policyUID, dnsInfo := range plugin.egressDNS.pdMap {
+			olddnsMap := dnsInfo.Get()
+			newdnsMap, err := dnsInfo.Update()
+			if err != nil {
+				glog.Error(err)
+				continue
+			}
+			if equal(olddnsMap, newdnsMap) {
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_dns.go]: equal() seems like it should be some operation on dnsInfo. Actually, since you don't even care about the contents of the maps here, you could just make dnsInfo.Update() return a bool (indicating whether anything changed) rather than a map (and not even call dnsInfo.Get())

File: pkg/sdn/plugin/egress_dns.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+			// found dns updates
+			for _, policy := range policies.Items {
+				if policy.UID == policyUID {
+					vnid, err := plugin.policy.GetVNID(policy.Namespace)
+					if err != nil {
+						glog.Warningf("Could not find netid for namespace %q: %v", policy.Namespace, err)
+						break
+					}
+					plugin.updateEgressNetworkPolicyRules(vnid)
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_dns.go]: Hm... plugin.updateEgressNetworkPolicyRules() is broken; it needs locking (eg, to protect against concurrent reads from/writes to plugin.egressPolicies[]). And then this will need to use the right locking as well.

File: pkg/sdn/plugin/egress_dns.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+					glog.Errorf("Invalid IP: %v for policy: %s", ip, policy.Name)
+				}
+			}
+			break
+		}
+	}
+	return cidrs
+}
+
+func (plugin *OsdnNode) SyncEgressDNSPolicyRules() {
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_dns.go]: I know we're already terrible about this, but I think it's confusing to keep spreading OsdnNode methods through more and more files. In this case, I think this method should go into egress_network_policy.go.

File: pkg/sdn/plugin/egress_dns.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+					}
+					plugin.updateEgressNetworkPolicyRules(vnid)
+					break
+				}
+			}
+		}
+	}
+}
+
+func (proxy *OsdnProxy) SyncEgressDNSProxyFirewall() {
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_dns.go]: and this should move to proxy.go...

File: pkg/sdn/plugin/egress_dns.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+				continue
+			}
+			if equal(olddnsMap, newdnsMap) {
+				continue
+			}
+
+			// found dns updates
+			for _, policy := range policies.Items {
+				if policy.UID == policyUID {
+					proxy.updateNetworkPolicy(policy)
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_dns.go]: As with the OVS case, this needs locking, though fortunately the remaining proxy code is already correct. So, you need to grab proxy.lock here.

And you have to call proxy.updateEndpoints() afterward too at the end of the loop if you changed anything.

**Comment by pravisankar** [comment]
@danwinship updated, ptal

**Comment by pravisankar** [comment]
[test]

**Comment by pravisankar** [comment]
Reiterating the concern that I raised during the standup:
In this PR, we are checking for dns updates every 30 mins. So in case of dns update, there is a period where egress policy may not behave as expected until it syncs these changes. Is this solution acceptable? If not:
(a) Do we want to call sync loop aggressively? (may not be ideal if we don't expect dns updates frequently) 
(b) Make this as OpenShift config option (similar to iptablesSyncPeriod)
(c) Make the code a bit more smart by querying TTL for the domain and poll every time when the TTL expires.

**Comment by danwinship** [comment]
> In this PR, we are checking for dns updates every 30 mins. So in case of dns update, there is a period where egress policy may not behave as expected until it syncs these changes. Is this solution acceptable?

I was going to say "yeah, I think most sites use a TTL of at least an hour normally anyway", but poking around, it seems that that's not true, especially for CDN-ed sites.

I don't think syncing more aggressively is a good solution though (either by default or as a config option). But also there's no easy way to get TTL... So I'm not sure what to do.

File: pkg/sdn/plugin/egress_dns.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (e *EgressDNS) GetNetCIDRs(policy osapi.EgressNetworkPolicy, dnsName string) []net.IPNet {
+	cidrs := []net.IPNet{}
+	for _, ip := range e.GetIPs(policy, dnsName) {
+		if ip.To4() != nil { // IPv4 addr
+			cidrs = append(cidrs, net.IPNet{IP: ip, Mask: net.CIDRMask(32, 32)})
+		} else if ip.To16() != nil { // IPv6 addr
+			// Currently we only handle IPv4 addrs
+			glog.Warningf("ignoring IPv6 addr: %s (domain: %s) for policy: %s", ip.String(), dnsName, policy.Name)
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_dns.go]: This (and the "else" below it) is dead code now; dnsInfo.Get() can only return valid IPv4 addresses

**Comment by knobunc** [comment]
I think we need to get the TTL... why is that hard?  Is it just that the library we are using can't provide it?

File: pkg/sdn/plugin/egress_network_policy.go [hunk 7]
```diff
@@ -25,12 +26,15 @@ func (plugin *OsdnNode) SetupEgressNetworkPolicy() error {
 			continue
 		}
 		plugin.egressPolicies[vnid] = append(plugin.egressPolicies[vnid], policy)
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_network_policy.go]: You need to grab the lock here (SetupEgressNetworkPolicy()) too. multiTenantPlugin.Start() calls vnids.Start() before node.SetupEgressNetworkPolicy(), and vnids.Start() spawns a goroutine which could theoretically end up calling UpdateEgressNetworkPolicyVNID().

File: pkg/sdn/plugin/egress_network_policy.go [hunk 8]
```diff
@@ -79,3 +92,44 @@ func (plugin *OsdnNode) UpdateEgressNetworkPolicyVNID(namespace string, oldVnid,
...
+		glog.V(5).Infof("periodic egress dns sync: update policy rules")
+
+		policies, err := plugin.osClient.EgressNetworkPolicies(kapi.NamespaceAll).List(kapi.ListOptions{})
+		if err != nil {
+			glog.Errorf("could not get EgressNetworkPolicies: %v", err)
+			continue
+		}
+
+		for _, policy := range policies.Items {
+			dnsInfo, ok := plugin.egressDNS.pdMap[policy.UID]
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_network_policy.go]: this isn't thread-safe is it?

File: pkg/sdn/plugin/node.go [hunk 9]
```diff
@@ -60,7 +61,11 @@ type OsdnNode struct {
 	kubeletInitReady   chan struct{}
 	iptablesSyncPeriod time.Duration
 	mtu                uint32
+
+	// Synchronizes operations on egressPolicies
+	egressPoliciesLock sync.Mutex
```
Comment (danwinship) [inline pkg/sdn/plugin/node.go]: (it would be nice to see this as a separate commit since it's fixing a separate pre-existing bug)

File: pkg/sdn/plugin/proxy.go [hunk 10]
```diff
@@ -149,3 +168,42 @@ EndpointLoop:
...
+				continue
+			}
+
+			if changed {
+				proxy.lock.Lock()
+				defer proxy.lock.Unlock()
+
+				proxy.updateNetworkPolicy(policy)
+				if proxy.allEndpoints != nil {
+					proxy.updateEndpoints()
```
Comment (danwinship) [inline pkg/sdn/plugin/proxy.go]: you want to do the updateEndpoints outside the `range policies.Items` loop

**Comment by danwinship** [comment]
> I think we need to get the TTL... why is that hard? Is it just that the library we are using can't provide it?

libc doesn't provide it, so in order to get it you need to write your own resolver basically

**Comment by knobunc** [comment]
I know calling the normal library functions doesn't return it... but is there no dns resolver library for golang already?

**Comment by danwinship** [comment]
There isn't even a good way to do it in C. Among other things, if you want to do it right, you need to parse /etc/nsswitch.conf and make sure you let /etc/hosts override DNS where it's supposed to, etc.

**Comment by pravisankar** [comment]
DNS updates are now based on domain TTL
@openshift/networking PTAL

**Comment by danwinship** [comment]
> There isn't even a good way to do it in C. Among other things, if you want to do it right, you need to parse /etc/nsswitch.conf and make sure you let /etc/hosts override DNS where it's supposed to, etc.

Just so that all the discussion is here in the PR: we decided that this doesn't really matter for our use case; this is primarily designed for dealing with site-external hostnames, so stuff like /etc/hosts and zeroconf names don't matter anyway.

File: pkg/sdn/plugin/egress_network_policy.go [hunk 11]
```diff
@@ -13,6 +14,10 @@ import (
 )
 
 func (plugin *OsdnNode) SetupEgressNetworkPolicy() error {
+	if err := CheckDNSResolver(); err != nil {
+		return err
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_network_policy.go]: Do we really want openshift to fail entirely if dig isn't present? (Rather than just making EgressNetworkPolicy not work?)

(Likewise in proxy.go)

File: pkg/sdn/plugin/egress_network_policy.go [hunk 12]
```diff
@@ -79,3 +91,57 @@ func (plugin *OsdnNode) UpdateEgressNetworkPolicyVNID(namespace string, oldVnid,
 		plugin.updateEgressNetworkPolicyRules(newVnid)
 	}
 }
+
+func (plugin *OsdnNode) syncEgressDNSPolicyRules() {
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_network_policy.go]: This is really complicated and confusing. You should make the EgressDNS code smarter and this code dumber. egress_network_policy.go shouldn't have to deal with the timers itself... EgressDNS should do that internally, and then only wake up egress_network_policy.go when something has actually changed. So this code would just be something like:

    func (plugin *OsdnNode) syncEgressDNSPolicyRules() {
            for {
                    policyNamespace <- plugin.egressDNS.changed
                    if vnid, err := plugin.policy.GetVNID(policyNamespace); err == nil {
                            plugin.updateEgressNetworkPolicyRules(vnid)
                    } else {
                            glog.Warningf("Could not find netid for namespace %q: %v", policyNamespace, err)
                    }
            }
    }

(Likewise in proxy.go)

File: pkg/sdn/plugin/egress_dns.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+	"github.com/golang/glog"
+
+	osapi "github.com/openshift/origin/pkg/sdn/api"
+
+	ktypes "k8s.io/kubernetes/pkg/types"
+)
+
+const (
+	defaultEgressDNSTTLSyncDuration = 8 * time.Hour
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_dns.go]: This gets defined here, but used from a different file in a different commit. I think you could just have the "syncEgressDNSTTL" goroutine here rather than in egress_network_policy.go though.

File: images/node/Dockerfile [hunk 14]
```diff
@@ -22,7 +22,7 @@ RUN curl -L -o /etc/yum.repos.d/origin-next-epel-7.repo https://copr.fedoraproje
     INSTALL_PKGS="libmnl libnetfilter_conntrack conntrack-tools openvswitch \
       libnfnetlink iptables iproute bridge-utils procps-ng ethtool socat openssl \
       binutils xz kmod-libs kmod sysvinit-tools device-mapper-libs dbus \
-      iscsi-initiator-utils" && \
+      iscsi-initiator-utils bind-utils" && \
```
Comment (danwinship) [inline images/node/Dockerfile]: Do we need similar changes anywhere else? dind images? RPM spec files?

File: pkg/sdn/plugin/dns.go [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+type DNS struct {
+	// Protects dnsMap operations
+	lock sync.Mutex
+	// Holds dns name and its corresponding information
+	dnsMap map[string]dnsValue
+}
+
+func CheckDNSResolver() error {
+	if _, err := exec.LookPath(DIG); err != nil {
+		return fmt.Errorf("%s is not installed", DIG)
```
Comment (pravisankar) [inline pkg/sdn/plugin/dns.go]: There are some go resolver libraries for DNS but I didn't find a go library that actually returns TTL value for a domain.

Comment (pravisankar) [inline pkg/sdn/plugin/dns.go]: If we just log an error instead of returning the error here then this feature will not work and I worry that the cluster admin may not notice this issue until someone tries this feature. I thought, throwing user error here will alert the admin to install the missing library.

File: pkg/sdn/plugin/dns.go [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+	res.ttl = ttl
+	d.dnsMap[dns] = res
+	return true
+}
+
+func (d *DNS) updateTTL(dns string) bool {
+	// Cached/Non-authoritative DNS servers only holds the remaining TTL for it to refresh but not the actual absolute TTL for the domain.
+	// Authoritative DNS servers holds actual TTL for the domain. So we need to:
+	// (a) Get all authorirative name servers for the given domain
+	// (b) Query TTL for the domain against one of the name servers from (a)
```
Comment (pravisankar) [inline pkg/sdn/plugin/dns.go]: We may not know exact TTL from non-auth server and may query few more times than required with this approach but yes, it definitely reduces the complexity.

File: pkg/sdn/plugin/dns.go [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+		if err != nil {
+			glog.Errorf("Failed to find authoritative name servers for domain: %q, err: %v, defaulting ttl=%s", dns, err, defaultTTL.String())
+		} else {
+			glog.Warningf("Could not find authoritative name servers for domain: %q, defaulting ttl=%s", dns, defaultTTL.String())
+		}
+		return d.setTTL(dns, defaultTTL)
+	}
+	nameServer := strings.Trim(string(out[:]), "\n")
+
+	out, err = exec.Command("sh", []string{"-c", fmt.Sprintf("echo $(%s | awk '{print $2}')", digShellCmdGetTTL(dns, nameServer))}...).Output()
```
Comment (pravisankar) [inline pkg/sdn/plugin/dns.go]: Good point. Right now, current approach is fetch TTL once from auth server, Update IPs for every TTL from non-auth server and another routine that updates TTL periodically from auth servers.

We could simply this to fetch and update TTL and IP from non-auth server and perform the above step again based on last obtained TTL. In this case, we don't need to contact auth server and no need of another routine to update TTL.

File: pkg/sdn/plugin/dns.go [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+	if err != nil || len(out) == 0 {
+		if err != nil {
+			glog.Errorf("Failed to get TTL value from authoritative name server for domain: %q, err: %v, defaulting ttl=%s", dns, err, defaultTTL.String())
+		} else {
+			glog.Warningf("TTL value not set by authoritative name server for domain: %q, defaulting ttl=%s", dns, defaultTTL.String())
```
Comment (pravisankar) [inline pkg/sdn/plugin/dns.go]: Yes, I think this shouldn't happen. I will change this to:
```
if err != nil || len(out) == 0 {
  	glog.Errorf("Failed to get TTL value from authoritative name server for domain: %q, err: %v, defaulting ttl=%s", dns, err, defaultTTL.String())
}
```

File: pkg/sdn/plugin/egress_network_policy.go (hunk 12 above)
Comment (pravisankar) [inline pkg/sdn/plugin/egress_network_policy.go]: Okay, I will make proxy and egress network policy register for dns changes and the go routine in egress dns that updates TTL and IP will inform all the registered callers when there is an IP update.

File: pkg/sdn/plugin/egress_dns.go (hunk 13 above)
Comment (pravisankar) [inline pkg/sdn/plugin/egress_dns.go]: This will be removed with the new approach.

File: images/node/Dockerfile (hunk 14 above)
Comment (pravisankar) [inline images/node/Dockerfile]: dind already has bind-utils package and may be I should also add to origin.spec? I will check

**Comment by pravisankar** [comment]
@danwinship Updated, PTAL

File: pkg/sdn/plugin/dns.go [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+			continue
+		}
+		if updated {
+			changed = true
+		}
+	}
+	return kerrors.NewAggregate(errList), changed
+}
+
+func (d *DNS) update(dns string) (error, bool) {
```
Comment (dcbw) [inline pkg/sdn/plugin/dns.go]: Maybe updateOne() to differentiate it from Update()?

File: pkg/sdn/plugin/dns.go [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+
+// Due to lack of any go bindings for dns resolver that actually provides TTL value, we are relying on 'dig' shell command.
+
+func digShellCmdGetIPAndTTL(dns string) string {
+	// Output Format:
+	// <domain-name>.		<<ttl from authoritative ns>	IN	A	<IP addr>
+	return fmt.Sprintf("%s +nocmd +noall +answer +ttlid a %s", DIG, dns)
+}
+
+func equal(oldips, newips []net.IP) bool {
```
Comment (dcbw) [inline pkg/sdn/plugin/dns.go]: I guess I'd name this ipsEqual() or something like that.

File: pkg/sdn/plugin/dns.go [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+	if len(oldips) != len(newips) {
+		return false
+	}
+
+	for _, oldip := range oldips {
```
Comment (dcbw) [inline pkg/sdn/plugin/dns.go]: An alternative would be to define an "ipsArray" type that implemented Sort (eg, Len(), Less(), and Swap()), sort at the end and do a direct compare.  Possibly more idiomatic but eh.

File: pkg/sdn/plugin/dns.go [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+		}
+		if updated {
+			changed = true
+		}
+	}
+	return kerrors.NewAggregate(errList), changed
+}
+
+func (d *DNS) update(dns string) (error, bool) {
+	out, err := exec.Command("sh", []string{"-c", digShellCmdGetIPAndTTL(dns)}...).Output()
```
Comment (dcbw) [inline pkg/sdn/plugin/dns.go]: Could we get some unit tests for this function?  Especially since it's parsing command output.

File: pkg/sdn/plugin/proxy.go [hunk 23]
```diff
@@ -49,15 +50,19 @@ func NewProxyPlugin(pluginName string, osClient *osclient.Client, kClient *kclie
...
+		osClient:  osClient,
+		ids:       make(map[string]uint32),
+		egressDNS: &EgressDNS{},
+		firewall:  make(map[string]*proxyFirewallItem),
 	}, nil
 }
 
 func (proxy *OsdnProxy) Start(baseHandler pconfig.EndpointsConfigHandler) error {
 	glog.Infof("Starting multitenant SDN proxy endpoint filter")
+	if err := CheckDNSResolver(); err != nil {
```
Comment (dcbw) [inline pkg/sdn/plugin/proxy.go]: Same question here as @danwinship asked in the previous commit; do we want to fail the node completely, rather than just make egress DNS not work?  This might catch people by surprise, especially when they don't use egress DNS at all.

**Comment by dcbw** [comment]
The only thing I worry about is multiple IPs that aren't returned in the same 'dig' invocation.  Like:

```
$ dig +nocmd +noall +answer +ttlid a google.com
google.com.		69	IN	A	216.58.218.142
$ dig +nocmd +noall +answer +ttlid a google.com
google.com.		169	IN	A	216.58.194.110
$ dig +nocmd +noall +answer +ttlid a google.com
google.com.		168	IN	A	216.58.194.110
$ dig +nocmd +noall +answer +ttlid a google.com
google.com.		65	IN	A	216.58.218.142
$ dig +nocmd +noall +answer +ttlid a google.com
google.com.		155	IN	A	216.58.194.110
$ dig +nocmd +noall +answer +ttlid a google.com
google.com.		116	IN	A	172.217.9.142
$ dig +nocmd +noall +answer +ttlid a www.google.com
www.google.com.		32	IN	A	216.58.194.132
```

Perhaps the egress policy gets one response, and the pod gets another.

**Comment by danwinship** [comment]
> The only thing I worry about is multiple IPs that aren't returned in the same 'dig' invocation.

Ugh... That kinda totally breaks this approach doesn't it?

**Comment by pravisankar** [comment]
@dcbw @danwinship Updated, PTAL
New changes:
- We drop all traffic and firewall rules for the namespace if DNS resolver, dig is not found on the node.
- Added test cases for add/update dns methods
- Minor updates based on review comments

**Comment by pravisankar** [comment]
[test]

**Comment by pravisankar** [comment]
updated docs pr: https://github.com/openshift/openshift-docs/pull/4092

File: pkg/sdn/plugin/dns_test.go [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+		expectTTLChange           bool
+		expectNextQueryTimeChange bool
+		expectFailure             bool
+	}
+
+	tests := []dnsTest{
+		{
+			testCase:                  "Test dns update of valid domain",
+			domainName:                "example.com",
+			dnsMapSize:                1,
```
Comment (danwinship) [inline pkg/sdn/plugin/dns_test.go]: Assuming that example.com will always only return 1 IP isn't safe.

I think a better way to do this would be to use kexec rather than os/exec in dns.go so you can test it against pre-recorded dig output. (eg, see pkg/util/ipcmd/ipcmd_test.go).

That can happen in a later commit (so this PR doesn't get delayed any more), but add a github issue or trello card to track it. Not sure if I think it would be better to commit this test as-is for now, or to just remove it from the PR for now.

File: pkg/sdn/plugin/egress_dns.go [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if dnsInfo, ok := e.pdMap[policyUID]; ok {
+		return dnsInfo.Update()
+	}
+	return nil, false
+}
+
+func (e *EgressDNS) SyncInit() {
```
Comment (danwinship) [inline pkg/sdn/plugin/egress_dns.go]: It would be more idiomatic to have a NewEgressDNS() method that returns a fully-initialized object, rather than using &EgressDNS{} from node/proxy and then having to call SyncInit()

File: pkg/sdn/plugin/egress_network_policy.go (hunk 11 above)
Comment (danwinship) [inline pkg/sdn/plugin/egress_network_policy.go]: You should just remove this from this commit rather than adding it here and then removing it in a later commit. (Likewise in proxy.go)

**Comment by pravisankar** [comment]
Updates:
- Now DNS tests use fake exec instead of making direct system calls
- Added NewEgressDNS(), NewDNS() methods to return full initialized objects
@openshift/networking PTAL

File: pkg/sdn/plugin/egress_dns.go [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+
+	timeSet := false
+	var minTime time.Time
+	var uid ktypes.UID
+
+	for policyUID, dnsInfo := range e.pdMap {
+		tm, ok := dnsInfo.GetMinQueryTime()
+		if !ok {
+			continue
+		}
```
Comment (pecameron) [inline pkg/sdn/plugin/egress_dns.go]: How long can e.lock be held here? Can this loop indefinite?

Comment (pravisankar) [inline pkg/sdn/plugin/egress_dns.go]: No, this loop is not indefinite. e.lock is held till we scan the egress policy dns map once.

**Comment by pravisankar** [comment]
flakes https://github.com/openshift/origin/issues/12072 https://github.com/openshift/origin/issues/10773 https://github.com/openshift/origin/issues/13271
re[test]

**Comment by danwinship** [comment]
LGTM

**Comment by pravisankar** [comment]
flaked again https://github.com/openshift/origin/issues/13318 
[test]

**Comment by knobunc** [comment]
[merge]

**Comment by pravisankar** [comment]
Resolved merge conflicts
[test]

**Comment by pravisankar** [comment]
```
[ERROR] PID 12674: hack/lib/build/environment.sh:137: `return "${exitcode}"` exited with status 2.
[INFO] 		Stack Trace: 
[INFO] 		  1: hack/lib/build/environment.sh:137: `return "${exitcode}"`
[INFO] 		  2: hack/lib/build/environment.sh:217: os::build::environment::withsource
[INFO] 		  3: hack/env:36: os::build::environment::run
[INFO]   Exiting with code 2.
++ export status=FAILURE
++ status=FAILURE
+ set +o xtrace
########## FINISHED STAGE: FAILURE: RUN CHECK AND VERIFY TASKS ##########
```
failure not related, looks like a flake 
re[test]

**Comment by danwinship** [comment]
> failure not related, looks like a flake
> re[ test ]

FYI (as the bot has already informed you), you can't just ask it to retest without a reason. If it actually is an infrastructure problem and there's no flake bug to link to, the easiest way around that is to rebase and repush.

But anyway, this wasn't a flake, it's just that the bot doesn't do a good job of highlighting the right things when the tests fail to compile:

    pkg/sdn/plugin/ovscontroller_test.go:542: not enough arguments in call to oc.UpdateEgressNetworkPolicyRules
    ...

**Comment by pravisankar** [comment]
[test]

**Comment by sghosh151** [comment]
Is this config supported on Atomic Host?

**Comment by pravisankar** [comment]
flake https://github.com/openshift/origin/issues/13068
re[test]
